src/libraries/System.Net.Sockets/src/System/Net/Sockets/UnixDomainSocketEndPoint.cs

```diff
@@ -10,13 +11,38 @@ namespace System.Net.Sockets
     /// <summary>Represents a Unix Domain Socket endpoint as a path.</summary>
     public sealed partial class UnixDomainSocketEndPoint : EndPoint
     {
+        // taken from System.IO.PathInternal
+        private static readonly StringComparison s_filePathComparison = OperatingSystem.IsWindows() ||
+                                                    OperatingSystem.IsMacOS() ||
+                                                    OperatingSystem.IsIOS() ||
+                                                    OperatingSystem.IsTvOS() ||
+                                                    OperatingSystem.IsWatchOS()
+                                                    ? StringComparison.OrdinalIgnoreCase
+                                                    : StringComparison.Ordinal;
+
         private const AddressFamily EndPointAddressFamily = AddressFamily.Unix;
 
         private readonly string _path;
         private readonly byte[] _encodedPath;
 
+        // The field is needed to distinguish the situation when
+        // _fullPath is not null but the object doesn't bound
+        // to the full path. _fullPath can be initialized by Equals/GetHashCode
+        // but it doesn't mean that the EndPoint is bound to the file system object.
+        private readonly bool _isBound;
+
+        // The field can be initialized lazily in the following circumstances:
+        // 1. Inside of Equals method
+        // 2. Inside of GetHashCode method
+        // 3. Inside of CreateBoundEndPoint method
+        // In case of non-abstract path, we need to have full path to the file system object.
+        // Otherwise, two endpoints may not be equal even if they pointing to the same file.
+        // Lazily initialized field then can be reused by these methods to avoid further
+        // allocations.
+        private string? _fullPath;
```

Comment:
Not actually new, the value of this field is used for two purposes. One of them is Equals/GetHashCode, yes. Otherwise, two paths pointing to the same location in fs will give different hash codes for EndPoint.

However, we can assume that relative paths pointing to the same location give us different EndPoints. If so, I can change the behavior.